File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Raytracer/Raytracer_CSharp/Sphere.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: Sphere.cs
//
//--------------------------------------------------------------------------

using System;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples
{
    class Sphere : SceneObject
    {
        public Vector Center;
        public double Radius;

        public Sphere(Vector center, double radius, Surface surface) : base(surface) { Center = center; Radius = radius; }

        public override ISect Intersect(Ray ray)
        {
            Vector eo = Vector.Minus(Center, ray.Start);
            double v = Vector.Dot(eo, ray.Dir);
            double dist;
            if (v < 0)
            {
                dist = 0;
            }
            else
            {
                double disc = Math.Pow(Radius, 2) - (Vector.Dot(eo, eo) - Math.Pow(v, 2));
                dist = disc < 0 ? 0 : v - Math.Sqrt(disc);
            }
            if (dist == 0) return ISect.Null;
            return new ISect(this, ray, dist);
        }

        public override Vector Normal(Vector pos)
        {
            return Vector.Norm(Vector.Minus(pos, Center));
        }
    }
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Raytracer/Raytracer_VisualBasic/Sphere.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: Sphere.vb
'
'--------------------------------------------------------------------------

Public Class Sphere
    Inherits SceneObject

    Public Center As Vector
    Public Radius As Double

    Public Sub New(ByVal center As Vector, ByVal radius As Double, ByVal surface As Surface)
        MyBase.New(surface)
        Me.Center = center
        Me.Radius = radius
    End Sub

    Public Overrides Function Intersect(ByVal ray As Ray) As ISect
        Dim eo = Vector.Minus(Center, ray.Start)
        Dim v = Vector.Dot(eo, ray.Direction)
        Dim dist As Double
        If v < 0 Then
            dist = 0
        Else
            Dim disc = Math.Pow(Radius, 2) - (Vector.Dot(eo, eo) - Math.Pow(v, 2))
            dist = If(disc < 0, 0, v - Math.Sqrt(disc))
        End If
        If dist = 0 Then Return Nothing
        Return New ISect(Me, ray, dist)
    End Function

    Public Overrides Function Normal(ByVal position As Vector) As Vector
        Return Vector.Norm(Vector.Minus(position, Center))
    End Function
End Class
